src/Transports.AspNetCore/Extensions/GraphQLBuilderUserContextExtensions.cs

```diff
@@ -80,6 +80,49 @@ public static IGraphQLBuilder AddUserContextBuilder<TUserContext>(this IGraphQLB
         return builder;
     }
 
+    /// <summary>
+    /// Configures a delegate to be used to create a user context for each GraphQL request.
+    /// <br/><br/>
+    /// Requires <see cref="IHttpContextAccessor"/> to be registered within the dependency injection framework
+    /// if calling <see cref="DocumentExecuter.ExecuteAsync(ExecutionOptions)"/> directly.
+    /// </summary>
+    public static IGraphQLBuilder AddUserContextBuilder<TUserContext>(this IGraphQLBuilder builder, Func<HttpContext, object?, TUserContext> creator)
+        where TUserContext : class, IDictionary<string, object?>
+    {
+        builder.Services.Register<IUserContextBuilder>(new UserContextBuilder<TUserContext>(creator ?? throw new ArgumentNullException(nameof(creator))));
+        builder.ConfigureExecutionOptions(options =>
+        {
+            if (options.UserContext == null || options.UserContext.Count == 0 && options.UserContext.GetType() == typeof(Dictionary<string, object>))
+            {
+                var requestServices = options.RequestServices ?? throw new MissingRequestServicesException();
+                var httpContext = requestServices.GetRequiredService<IHttpContextAccessor>().HttpContext!;
+                options.UserContext = creator(httpContext, null);
+            }
+        });
+
+        return builder;
+    }
+
+    /// <inheritdoc cref="AddUserContextBuilder{TUserContext}(IGraphQLBuilder, Func{HttpContext, TUserContext})"/>
+    public static IGraphQLBuilder AddUserContextBuilder<TUserContext>(this IGraphQLBuilder builder, Func<HttpContext, object?, Task<TUserContext>> creator)
+        where TUserContext : class, IDictionary<string, object?>
+    {
+        if (creator == null)
+            throw new ArgumentNullException(nameof(creator));
+        builder.Services.Register<IUserContextBuilder>(new UserContextBuilder<TUserContext>((context, payload) => new(creator(context, payload))));
+        builder.ConfigureExecutionOptions(async options =>
+        {
+            if (options.UserContext == null || options.UserContext.Count == 0 && options.UserContext.GetType() == typeof(Dictionary<string, object>))
```

Comment:
Why do we use `|| options.UserContext.Count == 0 && options.UserContext.GetType() == typeof(Dictionary<string, object>)` condition?

Reply:
Because the execution options are preinitialized with an empty dictionary by default. I think.

Reply:
Ah, and `IProvideUserContext` has only getter. Looks a bit awkward but OK.